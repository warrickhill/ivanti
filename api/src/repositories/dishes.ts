import { db } from "./db"
import { eq, getTableColumns, sql } from "drizzle-orm"
import { dishes, ratings } from "./schema"

type NewDish = typeof dishes.$inferInsert

export const getDishes = async () => {
    return await db
        .select({
            ...getTableColumns(dishes),
            rating: sql<number>`round(avg(${ratings.rating}), 1)`,
        })
        .from(dishes)
        .rightJoin(ratings, eq(dishes.id, ratings.dishId))
        .groupBy(dishes.id)
}

export const createDish = async (user: NewDish) => {
    return (await db.insert(dishes).values(user).returning())[0]
}

export const getDish = async (id: number) => {
    return (
        await db
            .select({
                ...getTableColumns(dishes),
                rating: sql<number>`round(avg(${ratings.rating}), 1)`,
            })
            .from(dishes)
            .rightJoin(ratings, eq(dishes.id, ratings.dishId))
            .groupBy(dishes.id)
            .where(eq(dishes.id, id))
    )[0]
}

export const updateDish = async (id: number, data: NewDish) => {
    return (
        await db.update(dishes).set(data).where(eq(dishes.id, id)).returning()
    )[0]
}

export const deleteDish = async (id: number) => {
    return await db.delete(dishes).where(eq(dishes.id, id))
}
