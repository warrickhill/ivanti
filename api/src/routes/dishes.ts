import { Elysia, t } from "elysia";
import reviews from "./reviews";

const dishSchema = t.Object({
  id: t.Optional(t.Integer()),
  name: t.String(),
  description: t.String(),
  image: t.String(),
  price: t.Number(),
});

export const dishes = new Elysia({ prefix: "/dishes" })
  .get("/", () => [{ test: "utrtrf" }])
  .post("/", ({ body }) => body, {
    body: dishSchema,
  })
  .get("/:id", () => {}, {
    params: t.Object({
      id: t.Number(),
    }),
  })
  .put("/:id", (body) => body, {
    body: dishSchema,
    params: t.Object({
      id: t.Number(),
    }),
  })
  .patch("/:id", (body) => body, {
    body: t.Partial(dishSchema),
    params: t.Object({
      id: t.Number(),
    }),
  })
  .delete("/:id", () => "hi", {
    params: t.Object({
      id: t.Number(),
    }),
  })
  .use(reviews)

export default dishes;
