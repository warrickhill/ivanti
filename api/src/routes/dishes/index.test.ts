// test/index.test.ts
import { describe, expect, it } from "bun:test"
import { Elysia } from "elysia"
import dishes from "."
import { edenTreaty } from "@elysiajs/eden"

const app = new Elysia().use(dishes).listen(3001)
const api = edenTreaty<typeof app>("http://localhost:3001")

describe("Dishes", () => {
    it("return an array of dishes", async () => {
        const { data } = await api.dishes.get()

        expect(data.length).toBe(96)
    })

    it("return an dish", async () => {
        const { data } = await api.dishes[1].get()

        expect(data).toMatchSnapshot()
    })

    it("creates, updates and deletes a dish", async () => {
        const { data } = await api.dishes.post({
            name: "Test Dish",
            description: `The name name says it all`,
            image: "https://fakeimg.pl/600x400",
            price: 10.99,
        })
        expect(data?.name).toBe("Test Dish")

        const { data: updated } = await api.dishes[data.id].patch({
            name: "Updated test dish",
        })

        expect(updated?.name).toBe("Updated test dish")

        const { data: deleted } = await api.dishes[data.id].delete()

        expect(deleted.lastInsertRowid).toBe(data.id)
    })
})
