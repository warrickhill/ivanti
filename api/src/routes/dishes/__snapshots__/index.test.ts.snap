// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Dishes return an dish 1`] = `
{
  "description": "The name Chicken and Barbecue sauce with Rice says it all",
  "id": 1,
  "image": "https://fakeimg.pl/600x400",
  "name": "Chicken and Barbecue sauce with Rice",
  "price": 10.99,
  "rating": 3,
}
`;
